src/Components/Web.JS/src/Boot.WebAssembly.ts

```diff
@@ -73,6 +76,10 @@ async function boot(options?: Partial<WebAssemblyStartOptions>): Promise<void> {
     }
   };
 
+  if (document.getElementById('blazor-default-loading')) {
+    WebAssemblyProgressReporter.init();
+  }
```

Comment:
Hmm, I wonder if we need to have some mechanism here to initialize a custom progress reporter. Users can't do it before `Blazor.start` because the `Blazor.webAssemblyProgressService` wouldn't be available then and doing it after `Blazor.start` would be too late as the resources would've already been loaded.

Maybe we need to have a `Blazor.webAssemblyProgressReporter` within `GlobalExports`, and if the user hasn't preset that, then we use our default `WebAssemblyProgressReporter`. 

What do you think about this, am I missing something?

Reply:
What would be the difference between adding `Blazor.webAssemblyProgressReporter` in `GlobalExports` versus the current implementation that we have? Is this change needed in order for users to have their own custom progress reporter, if they do not want to use the default?

Reply:
With the current implementation, if a user wants to use their own `ProgressReporter`, they don't have a way to `attach` their custom `ProgressReporter` to the `webAssemblyProgressService`. Users can't do it before `Blazor.start` because the `Blazor.webAssemblyProgressService` wouldn't be available then and doing it after `Blazor.start` would be too late as the resources would've already been loaded.

By adding a `Blazor.webAssemblyProgressReporter` in `GlobalExports`, users can set a custom `ProgressReporter` should they desire, and we'd handle calling `Blazor.webAssemblyProgressReporter.init()` if `Blazor.webAssemblyProgressReporter` exists. Given that, we may also want to define a `IBlazorWebAssemblyProgressReporter` interface with `init` and `setProgress` methods (though this is presently optional as we don't publish Blazor Typescript Definitions).

Reply:
Given we have a `Blazor.webAssemblyProgressReporter`, we can also move the `Blazor.webAssemblyProgressService` to [`Blazor._internal.webAssemblyProgressService`](https://github.com/dotnet/aspnetcore/blob/835dc9acc868c25794dc3bddbc858bd9da24bd14/src/Components/Web.JS/src/GlobalExports.ts#L32), and just have the `Blazor.webAssemblyProgressReporter` as the public API for custom progress reporters. We had `Blazor.webAssemblyProgressService` in `GlobalExports` to make `attach`ing custom progress reporters easier. But with  `Blazor.webAssemblyProgressReporter` being in `GlobalExports`, that should no longer be a concern. We can then just pass in the `Blazor._internal.webAssemblyProgressService.attach` function to the `Blazor.webAssemblyProgressReporter.init` so the progress reporter can attach it's own `setProgress` function as required.

Would that make sense? What're your thoughts on something like this?